### DISCUSSION TITLE
How to block Subscribers from Publishing and Playing the streams?

### DISCUSSION THREAD

**Discussion by Mohit-3196** [ant-media/Ant-Media-Server#5796] [General]
Hello,
I have a use case where I want users to publish/play the streams for specific durations according to their subscription plan and therefore would like to stop users from publishing and playing streams according to the allotted time.

It seems that the feature to block subscribers from publishing and playing is now available in v2.7.0 as per the release logs [https://github.com/ant-media/Ant-Media-Server/releases](url)

Can anyone share with me the steps/configurations needed to achieve this? Also, for which publish and play protocol does this feature apply?

Thank you.

**Comment by lastpeony** [comment]
The subscriber block feature allows blocking a specific user from engaging in playback, publishing, or both at any given moment. This implies that even if the user is actively publishing or playing the stream, their ability to publish or play will cease until the block is removed or expires. Block is valid for all publish and play types.

To utilize subscriber block for play first enable TOTP(Time-based One-Time Password) for play through web panel application settings.
![image](https://github.com/ant-media/Ant-Media-Server/assets/24750845/481fb87d-69de-4487-bbf9-19437fd44448)
If you want rather than enabling it from UI you can also add 
`timeTokenSubscriberOnly=true
` and 
`timeTokenSecretForPlay="O24FW6"
`through app settings, advanced section.


By default, the TOTP generated for playback remains valid for 60 seconds after its generation. Consequently, users intending to utilize this token must send a play request to AMS within this 60-second timeframe.
You can change default TOTP time by adding
`timeTokenPeriod=60`
to app settings.

When this setting is enabled users without TOTP wont be able to watch the streams. 

Now you can send a `GET ` request to generate TOTP for the subscriber. 
`http://localhost:5080/WebRTCAppEE/rest/v2/broadcasts/{streamId}/subscribers/{subscriberId}/totp`
Add "type":"play" as query param to get request.

![image](https://github.com/ant-media/Ant-Media-Server/assets/24750845/62748c64-deaa-4dc1-898f-d1e897a3c6fa)

Example curl:
`curl --location 'http://localhost:5080/LiveApp/rest/v2/broadcasts/teststream/subscribers/lastpeony/totp?type=play'
`
You can retrieve the token from `dataId `field of the returned object.

As an illustration, suppose you are associating your users with userIds in your application. When users initiate playback on your application, you can transmit their userId as the subscriberId and issue a TOTP generation request to AMS. Upon receiving the token, pass it to the Ant Media Server SDK, and commence the user's playback.

Users must pass TOTP as `subscriberCode ` to ant media server from SDK when they send play request.

`http://localhost:5080/LiveApp/play.html?id=teststream&subscriberId=lastpeony&subscriberCode=956364
`
Upon a user successfully initiating a play request with TOTP, they become authenticated and gain access to watch the stream. It's important to note that they won't be able to play the stream if they refresh the page and their TOTP has expired. However, if the TOTP is still valid and they refresh, they will be reauthenticated and able to resume playing the stream.


Now, you can utilize the subscriber block feature to prevent this user from playing for the desired duration.

To do that send a `PUT `request to
http://localhost:5080/LiveApp/rest/v2/broadcasts/{streamId}/subscribers/block/{blockDurationInSeconds}/{blockType}

blockType can be "play","publish" or "publish_play"

Example request which will block subscriber lastpeony play for 120 seconds:
`http://localhost:5080/LiveApp/rest/v2/broadcasts/teststream/subscribers/lastpeony/block/120/play
`
![image](https://github.com/ant-media/Ant-Media-Server/assets/24750845/230fa2bc-8220-4918-9e38-19cc7164444c)

Example curl:
`curl --location --request PUT 'http://localhost:5080/LiveApp/rest/v2/broadcasts/teststream/subscribers/lastpeony/block/120/play'`
`
Observe that players playback immediately stops when this request returns success.

After 120 seconds user will be able to play the stream again. To remove the block manually you can pass 0 to the block request.
Example curl:
`curl --location --request PUT 'http://localhost:5080/LiveApp/rest/v2/broadcasts/teststream/subscribers/lastpeony/block/0/play'`

Please notify that playback resumes immediately upon the successful return of this request.

For publish process is same with play.
First enable TOTP for publishing through web panel. Also generate a secret for it.
![image](https://github.com/ant-media/Ant-Media-Server/assets/24750845/10799b22-9f87-4c23-875f-81e36cf25daf)
Just like play scenario if you want rather than enabling it from UI you can also add 
`timeTokenSubscriberOnly=true
` and 
`timeTokenSecretForPublish="4ULgui"
`through app settings, advanced section.

Dont forget to save application settings.

Now you can send a `GET ` request to generate TOTP for the subscriber. 
`http://localhost:5080/WebRTCAppEE/rest/v2/broadcasts/{streamId}/subscribers/{subscriberId}/totp`
Add "type":"publish" as query param to get request.

`http://localhost:5080/LiveApp/rest/v2/broadcasts/teststream/subscribers/lastpeony/totp?type=publish
`.
![image](https://github.com/ant-media/Ant-Media-Server/assets/24750845/7e40337e-2144-46db-8a67-93ea778f2a29)

Example curl:
`curl --location 'http://localhost:5080/LiveApp/rest/v2/broadcasts/teststream/subscribers/lastpeony/totp?type=publish'
`
Pass the obtained TOTP token as `secretCode` in publish requests to Ant Media Server SDKs or in RTMP publish requests.

For example on javascript SDK you should call publish command as follows:
`webRTCAdaptor.publish(streamId, null, subscriberId, subscriberCode);
`
Example:
`webRTCAdaptor.publish("teststream", null, "lastpeony", "451222")`
(2nd parameter which is null here is token)

Similarly on RTMP publishing you need to pass parameters streamId, subscriberId and subscriberCode.
OBS configuration should look like this:
![image](https://github.com/ant-media/Ant-Media-Server/assets/24750845/7574e20d-b89f-4b45-940b-a0a7494efda4)
Example ffmpeg command to publish with RTMP:

`ffmpeg -re -i input_source -c:v libx264 -c:a aac -f flv "rtmp://172.17.52.70/LiveApp/teststream?subscriberId=lastpeony&subscriberCode=451222"`

Since now you are publishing using subscriber TOTP, you can block this subscriber from publishing with block request.
To block the user from publishing for 120 seconds send subscriber block `PUT `request to:
`http://localhost:5080/LiveApp/rest/v2/broadcasts/teststream/subscribers/lastpeony/block/120/publish

![image](https://github.com/ant-media/Ant-Media-Server/assets/24750845/08121a93-3888-4a30-b898-737a716108a1)
Example curl:
`curl --location --request PUT 'http://localhost:5080/LiveApp/rest/v2/broadcasts/teststream/subscribers/lastpeony/block/120/publish'
`
As soon as this request returns success publishing of the subscriber will stop and they will be blocked for 120 seconds.

To remove the block pass 0 seconds to request.
Example curl:
`curl --location --request PUT 'http://localhost:5080/LiveApp/rest/v2/broadcasts/teststream/subscribers/lastpeony/block/0/publish'`

Keep in mind that if you block subscribers from publishing and then unblock them, they will encounter an "unauthorized_access" error if their TOTP has expired. Consequently, they won't be able to publish even if the block has been lifted. In such instances, you'll need to generate a new TOTP.
